File: HoverDome_OC/HoverDome_OC/ViewController.m


#import "ViewController.h"
#import "HoverPageViewController.h"
#import "Children1ViewController.h"
#import "Children2ViewController.h"
#import "Children3ViewController.h"
@interface ViewController ()<HoverPageViewControllerDelegate>
@property(nonatomic, strong) UIView *indicator;
@property(nonatomic, strong) HoverPageViewController *hoverPageViewController;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"悬停";
    
    UILabel *headerView = [UILabel new];
    headerView.frame = CGRectMake(0, 0, self.view.frame.size.width, 200);
    headerView.text = @"可上下滑动头部";
    headerView.textColor = [UIColor whiteColor];
    headerView.backgroundColor = [UIColor  redColor];
    headerView.textAlignment = NSTextAlignmentCenter;
    
    /// 指示器
    UIView *pageTitleView = [UIView new];
    pageTitleView.backgroundColor = [UIColor lightGrayColor];
    pageTitleView.frame = CGRectMake(0, 0, self.view.frame.size.width, 40);
    
    /// 添加3个按钮
    CGSize buttonSize = CGSizeMake(self.view.frame.size.width / 3, pageTitleView.frame.size.height);
    for (int i = 0; i < 3; i++) {
        UIButton *button = [UIButton new];
        button.tag = i;
        [button addTarget:self action:@selector(buttonClick:) forControlEvents:UIControlEventTouchUpInside];
        button.frame = CGRectMake(i * buttonSize.width, 0, buttonSize.width, buttonSize.height);
        [button setTitle:@"控制器" forState:UIControlStateNormal];
        [button setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        [pageTitleView addSubview:button];
    }
    
    UIButton *button = pageTitleView.subviews.firstObject;
    [button layoutIfNeeded];
    
    /// 指示器底部
    self.indicator = [UIView new];
    self.indicator.frame = CGRectMake(0, buttonSize.height - 3, buttonSize.width, 3);
    self.indicator.backgroundColor = [UIColor yellowColor];
    [pageTitleView addSubview: self.indicator];
    
    /// 添加子控制器
    NSMutableArray *viewControllers = [NSMutableArray array];
    HoverChildViewController *vc1 = [[Children1ViewController alloc]init];
    HoverChildViewController *vc2 = [[Children2ViewController alloc]init];
    HoverChildViewController *vc3 = [[Children3ViewController alloc]init];
    [viewControllers addObject:vc1];
    [viewControllers addObject:vc2];
    [viewControllers addObject:vc3];
    
    /// 计算导航栏高度
    CGFloat barHeight = [UIApplication sharedApplication].statusBarFrame.size.height + self.navigationController.navigationBar.frame.size.height;

     /// 添加分页控制器
    self.hoverPageViewController = [HoverPageViewController viewControllers:viewControllers headerView:headerView pageTitleView:pageTitleView];
    self.hoverPageViewController.view.frame = CGRectMake(0, barHeight, self.view.frame.size.width, self.view.frame.size.height - barHeight);
    self.hoverPageViewController.delegate = self;
    [self addChildViewController:self.hoverPageViewController];
    [self.view addSubview:self.hoverPageViewController.view];
}

- (void)buttonClick:(UIButton *)btn{
    [self.hoverPageViewController moveToAtIndex:btn.tag animated:YES];
}

- (void)hoverPageViewController:(HoverPageViewController *)ViewController scrollViewDidScroll:(UIScrollView *)scrollView{
    CGFloat progress = scrollView.contentOffset.x / scrollView.frame.size.width;
    self.indicator.frame = CGRectMake(self.indicator.frame.size.width * progress, self.indicator.frame.origin.y, self.indicator.frame.size.width, self.indicator.frame.size.height);

}

@end

File: HoverDome/HoverDome/ViewController.swift
//
//  ViewController.swift
//  HoverDome
//
//  Created by admin on 2019/2/27.
//  Copyright © 2019 zhouqiao. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    var hoverPageViewController:HoverPageViewController!
    let indicator = UIView()
    var indicatorMargin:CGFloat = 0
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white
        title = "悬停"
        prepareView()
    }
}

extension ViewController {

    private func prepareView() {
        let headerView = UILabel()
        headerView.frame.size = CGSize(width: view.frame.width, height: 200)
        headerView.backgroundColor = UIColor.red
        headerView.text = "可上下滑动头部"
        headerView.textColor = UIColor.white
        headerView.textAlignment = .center
        
        /// 指示器
        let pageTitleView = UIView()
        pageTitleView.frame.size = CGSize(width: view.frame.width, height: 40)

        /// 添加3个按钮
        let buttonSize = CGSize(width: view.frame.width / 3, height: pageTitleView.frame.height)
        for i in 0..<3 {
            let button = UIButton()
            button.tag = i
            button.frame.size = buttonSize
            button.backgroundColor = UIColor.lightGray
            button.frame.origin.x = CGFloat(i) * buttonSize.width
            button.setTitleColor(UIColor.black, for: .normal)
            button.setTitle("控制器", for: .normal)
            button.addTarget(self, action: #selector(ViewController.buttonClick), for: .touchUpInside)
            pageTitleView.addSubview(button)
        }
        
        let button = pageTitleView.subviews[0] as! UIButton
        button.layoutIfNeeded()
        indicator.frame.size = CGSize(width: (button.titleLabel?.frame.width)!, height: 3)
        indicator.backgroundColor = UIColor.yellow
        indicator.center.x = button.center.x
        indicator.frame.origin.y = button.frame.height - indicator.frame.height
        indicatorMargin = indicator.frame.origin.x
        pageTitleView.addSubview(indicator)
        
        /// 添加子控制器
        var viewControllers = [HoverChildViewController]()
        let vc1 = Children1ViewController()
        let vc2 = Children2ViewController()
        let vc3 = Children3ViewController()
        viewControllers.append(vc1)
        viewControllers.append(vc2)
        viewControllers.append(vc3)
        
        /// 计算导航栏高度
        var barHeight = UIApplication.shared.statusBarFrame.height
        if let bar = navigationController?.navigationBar{
            barHeight+=bar.frame.height
        }
        
        /// 添加分页控制器 
        hoverPageViewController = HoverPageViewController(viewControllers: viewControllers, headerView: headerView, pageTitleView: pageTitleView)
        hoverPageViewController.delegate = self
        hoverPageViewController.view.frame = CGRect(x: 0, y: barHeight, width: view.frame.width, height: view.frame.height - barHeight)
        addChild(hoverPageViewController)
        view.addSubview(hoverPageViewController.view)
    }

    @objc func buttonClick(btn: UIButton) {
        hoverPageViewController.move(to: btn.tag, animated: true)
    }
}

extension ViewController:HoverPageViewControllerDelegate{
    
    func hoverPageViewController(_ viewController: HoverPageViewController, scrollViewDidScroll scrollView: UIScrollView) {
        let progress = scrollView.contentOffset.x / scrollView.frame.width
        indicator.frame.origin.x = ((indicator.frame.width + (indicatorMargin * 2)) * progress) + indicatorMargin
    }
}

